File: DispatchMain.m
//
//  DispatchAsyncMain.m
//
//  Created by Nick Brook on 16/12/2016.
//  Copyright © 2016 Nick Brook. All rights reserved.
//

#import "DispatchMain.h"

BOOL dispatch_async_main_if_required(dispatch_block_t block) {
    if([NSThread isMainThread]) {
        block();
        return NO;
    }
    dispatch_async(dispatch_get_main_queue(), block);
    return YES;
}

BOOL dispatch_sync_main_if_required(dispatch_block_t block) {
    if([NSThread isMainThread]) {
        block();
        return NO;
    }
    dispatch_sync(dispatch_get_main_queue(), block);
    return YES;
}

File: DispatchMain.swift
//
//  DispatchMain.swift
//
//  Created by Nick Brook on 06/03/2017.
//  Copyright © 2017 Nick Brook. All rights reserved.
//

import Foundation

public extension DispatchQueue {
    public static func asyncMainIfRequired(group: DispatchGroup? = nil, qos: DispatchQoS = DispatchQoS.default, flags: DispatchWorkItemFlags = [], execute work: @escaping @convention(block) () -> Swift.Void) {
        guard Thread.isMainThread else {
            DispatchQueue.main.async(group: group, qos: qos, flags: flags, execute: work)
            return
        }
        work()
    }
    
    public static func syncMainIfRequired(execute block: () -> Swift.Void) {
        guard Thread.isMainThread else {
            DispatchQueue.main.sync(execute: block)
            return
        }
        block()
    }
}
